File: src/main/java/com/flexpoker/game/command/handlers/JoinGameCommandHandler.java
package com.flexpoker.game.command.handlers;

import javax.inject.Inject;

import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import com.flexpoker.framework.command.CommandHandler;
import com.flexpoker.framework.event.EventPublisher;
import com.flexpoker.game.command.commands.JoinGameCommand;
import com.flexpoker.game.command.factory.GameFactory;
import com.flexpoker.game.command.framework.GameEvent;
import com.flexpoker.game.command.repository.GameEventRepository;

@Component
public class JoinGameCommandHandler implements CommandHandler<JoinGameCommand> {

    private final GameFactory gameFactory;

    private final EventPublisher<GameEvent> eventPublisher;

    private final GameEventRepository gameEventRepository;

    @Inject
    public JoinGameCommandHandler(GameFactory gameFactory,
            EventPublisher<GameEvent> eventPublisher,
            GameEventRepository gameEventRepository) {
        this.gameFactory = gameFactory;
        this.eventPublisher = eventPublisher;
        this.gameEventRepository = gameEventRepository;
    }

    @Async
    @Override
    public void handle(JoinGameCommand command) {
        var gameEvents = gameEventRepository.fetchAll(command.getAggregateId());
        var game = gameFactory.createFrom(gameEvents);
        game.joinGame(command.getPlayerId());
        var eventsWithVersions = gameEventRepository.setEventVersionsAndSave(gameEvents.size(), game.fetchNewEvents());
        eventsWithVersions.forEach(x -> eventPublisher.publish(x));
    }

}

File: src/main/kotlin/com/flexpoker/game/command/handlers/JoinGameCommandHandler.kt
package com.flexpoker.game.command.handlers

import com.flexpoker.framework.command.CommandHandler
import com.flexpoker.framework.event.EventPublisher
import com.flexpoker.game.command.aggregate.applyEvents
import com.flexpoker.game.command.aggregate.eventproducers.joinGame
import com.flexpoker.game.command.commands.JoinGameCommand
import com.flexpoker.game.command.events.GameEvent
import com.flexpoker.game.command.repository.GameEventRepository
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class JoinGameCommandHandler @Inject constructor(
    private val eventPublisher: EventPublisher<GameEvent>,
    private val gameEventRepository: GameEventRepository
) : CommandHandler<JoinGameCommand> {

    @Async
    override fun handle(command: JoinGameCommand) {
        val gameEvents = gameEventRepository.fetchAll(command.aggregateId)
        val state = applyEvents(gameEvents)
        val newEvents = joinGame(state, command.playerId)
        val eventsWithVersions = gameEventRepository.setEventVersionsAndSave(gameEvents.size, newEvents)
        eventsWithVersions.forEach { eventPublisher.publish(it) }
    }

}